### PR TITLE
Refactor/Fix sponge function

### LINKED ISSUE #50: Allign calling convention of `zelbet` and `sponge`
**Describe what you want implemented** 
`zelbet` takes an array and `sponge` takes a vector but it would make sense to have both functions have the same calling convention.
This could be a reference, a mutable reference or an array. Investigate what makes most sense.

**Describe "Why" this is needed**
for ease of use

**Describe alternatives you've considered**
n/a

**Additional context**
n/a

### PR DESCRIPTION
**Pull request by moCello**
I used the change in the function signatures for issue #50 to also refactor sponge internally to make it easier to follow.

The capacity of the state was initialized with a one but it must be a zero -> Fixed

We used to have sponge return an output vector of variable length. However in our use case this output vector will always be one element long. We therefore decided to remove complexity and get rid of the `output_len` parameter. We will always return one `BlsScalar`.
See also the discussion bellow

### CODE REVIEW

File: src/sponge.rs [hunk 1]
```diff
@@ -8,131 +8,120 @@
...
-};
+use crate::permutation;
 use dusk_plonk::prelude::{BlsScalar, Constraint, TurboComposer, Witness};
 
 /// Zelbet sponge function in hash version, operating on `BlsScalars`.
-pub fn hash(input: Vec<BlsScalar>, output_len: usize) -> Vec<BlsScalar> {
-    // Initialize input as mutable
-    let mut input = input;
-
+pub fn hash(input: &[BlsScalar], output_len: usize) -> Vec<BlsScalar> {
```
Comment (vlopes11) [inline src/sponge.rs]: There is something off with this function. Normally we use a single bit digest as output:

`pub fn hash(input: &[BlsScalar]) -> BlsScalar`

Here we are mixing functionalities. You are creating an iterator that could squeeze `n` elements, while also providing a normal digest function.

I suggest you split this function in two:

`pub fn digest(input: &[BlsScalar]) -> BlsScalar`
`pub fn stream(input: &[BlsScalar]) -> impl Iterator<Item = BlsScalar>`

Comment (vlopes11) [inline src/sponge.rs]: There is something off with this function. Normally we use a single bit digest as output:

`pub fn hash(input: &[BlsScalar]) -> BlsScalar`

Here we are mixing functionalities. You are creating an iterator that could squeeze `n` elements, while also providing a normal digest function.

I suggest you split this function in two:

`pub fn digest(input: &[BlsScalar]) -> BlsScalar`
`pub fn stream(input: &[BlsScalar]) -> impl Iterator<Item = BlsScalar>`

**Review by vlopes11 — CHANGES_REQUESTED** [review]

File: src/sponge.rs (hunk 1 above)
Comment (vlopes11) [reply src/sponge.rs]: For that, you will need some iterator wrapper struct that will contain a state and its used element

Comment (moCello) [reply src/sponge.rs]: Not sure I understand what you mean. I follow the sponge framework that is divided in three phases:
- Pad `input`
- Absorb `input` into `state`
- Squeeze `state` to generate `output` of the given length `output_len`

I could of course create three different functions for those phases but I don't really see what is being achieved with that. But maybe I am missing your point.

Comment (moCello) [reply src/sponge.rs]: We talked earlier about the output of sponge. It actually can be of different length which is why we have `output_len` as a parameter. If we decide however that we want ours to always return one `BlsScalar` I could of course change that.

Comment (vlopes11) [reply src/sponge.rs]: We are talking specifically about the `squeeze` phase. The `pad` and `absorb` is fine

Yes, the output might be variable size. However, many use cases will only get one field element; so its a waste to create a vector to store a single element.

As for the variable size, if we are creating a vector, we are eager loading and allocating unnecessarily. If its an iterator, the consumer can decide what to do with that (including creating a vector with `collect::<Vec<BlsScalar>>`

Comment (moCello) [reply src/sponge.rs]: Ok, I see. I think you are right then and the easiest would be to simply return one `BlsScalar` so we don't introduce complexity that we don't use anyway.

Comment (vlopes11) [reply src/sponge.rs]: We might use multiple outputs of the sponge tho. We could leave an open issue to implement the iterator in the future

Comment (moCello) [reply src/sponge.rs]: Sure! Although given the fact that we didn't need sponge to return multiple outputs up until this point, I find it unlikely that we will need it in the future. But one never knows.

File: src/sponge.rs [hunk 2]
```diff
@@ -4,135 +4,65 @@
...
 
 /// Zelbet sponge function in hash version, operating on `BlsScalars`.
-pub fn hash(input: Vec<BlsScalar>, output_len: usize) -> Vec<BlsScalar> {
-    // Initialize input as mutable
-    let mut input = input;
-
+pub fn hash(input: &[BlsScalar]) -> BlsScalar {
     // Pad input
+    // After this, input_len is a multiple of 2
+    let mut input = input.to_vec();
```
Comment (vlopes11) [inline src/sponge.rs]: As improvement for the future (or now), this can easily be an iterator instead of a vector

File: tests/sponge.rs [hunk 3]
```diff
@@ -12,9 +12,9 @@ use zelbet::sponge;
 
 fn sponge_circuit(composer: &mut TurboComposer, input: BlsScalar) {
```
Comment (vlopes11) [inline tests/sponge.rs]: Also for the future, but it doesn't really make sense to test a sponge function with a single scalar. This test should cover edge cases such as 0, 1, 2 and many scalars as input

**Review by vlopes11 — APPROVED** [review]
LGTM

File: tests/sponge.rs (hunk 3 above)
Comment (moCello) [reply tests/sponge.rs]: Good point! I will add more test cases

File: src/sponge.rs (hunk 2 above)
Comment (moCello) [reply src/sponge.rs]: You are right, I'll change it now

**Review by ureeves — APPROVED** [review]
LGTM

**Review by vlopes11 — APPROVED** [review]
LGTM
